src/java.base/share/classes/jdk/internal/lang/stable/StableValueImpl.java

```diff
@@ -0,0 +1,218 @@
+            return false;
+        }
+        // Prevent reentry via an orElseSet(supplier)
+        preventReentry();
+        // Mutual exclusion is required here as `orElseSet` might also
+        // attempt to modify the `wrappedValue`
+        synchronized (this) {
+            return wrapAndSet(contents);
+        }
+    }
+
+    @ForceInline
+    @Override
+    public void setOrThrow(T contents) {
+        if (!trySet(contents)) {
+            // Neither the set contents nor the provided contents is revealed in the
+            // exception message as it might be sensitive.
+            throw new IllegalStateException("The contents is already set");
+        }
+    }
+
+    @ForceInline
+    @Override
+    public T orElseThrow() {
+        final Object t = wrappedContentAcquire();
+        if (t == null) {
+            throw new NoSuchElementException("No contents set");
+        }
+        return unwrap(t);
+    }
+
+    @ForceInline
+    @Override
+    public T orElse(T other) {
+        final Object t = wrappedContentAcquire();
+        return (t == null) ? other : unwrap(t);
+    }
+
+    @ForceInline
+    @Override
+    public boolean isSet() {
+        return wrappedContentAcquire() != null;
+    }
+
+    @ForceInline
+    @Override
+    public T orElseSet(Supplier<? extends T> supplier) {
+        Objects.requireNonNull(supplier);
+        final Object t = wrappedContentAcquire();
+        return (t == null) ? orElseSetSlowPath(supplier) : unwrap(t);
+    }
+
+    @DontInline
+    private T orElseSetSlowPath(Supplier<? extends T> supplier) {
+        preventReentry();
+        synchronized (this) {
+            final Object t = contents;  // Plain semantics suffice here
+            if (t == null) {
+                final T newValue = supplier.get();
+                // The mutex is not reentrant so we know newValue should be returned
+                wrapAndSet(newValue);
+                return newValue;
+            }
+            return unwrap(t);
+        }
+    }
+
+    // The methods equals() and hashCode() should be based on identity (defaults from Object)
+
+    @Override
+    public String toString() {
+        final Object t = wrappedContentAcquire();
+        return t == this
+                ? "(this StableValue)"
+                : renderWrapped(t);
+    }
+
+    // Internal methods shared with other internal classes
+
+    @ForceInline
+    public Object wrappedContentAcquire() {
+        return UNSAFE.getReferenceAcquire(this, CONTENT_OFFSET);
+    }
+
+    static String renderWrapped(Object t) {
+        return (t == null) ? UNSET_LABEL : Objects.toString(unwrap(t));
+    }
+
+    // Private methods
+
+    // This method is not annotated with @ForceInline as it is always called
+    // in a slow path.
+    private void preventReentry() {
+        if (Thread.holdsLock(this)) {
+            throw new IllegalStateException("Recursive initialization of a stable value is illegal");
```

Comment:
I don't know how common it is to use holdsLocks for control flow but just to say that there are several places in the JDK that have to detect reentrancy and some of these are good candidates to use ScopedValues.